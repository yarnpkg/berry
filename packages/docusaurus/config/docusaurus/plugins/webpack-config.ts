import type {Options as MDXLoaderOptions, MDXOptions} from '@docusaurus/mdx-loader';
import type {LoadContext, Plugin}                     from '@docusaurus/types';
import {createRequire}                                from 'node:module';
import path                                           from 'node:path';

const webpack = createRequire(require.resolve(`@docusaurus/core/package.json`))(`webpack`);

export type Options = {
  changelog: Partial<MDXOptions>;
};

const plugin = async function(context: LoadContext, options: Options): Promise<Plugin> {
  return {
    name: `docusaurus-plugin-yarn-webpack-config`,
    configureWebpack(config, isServer, utils) {
      return {
        module: {
          rules: [
            {
              test: /\.term\.dat$/,
              use: [require.resolve(`../../webpack/ansi-loader.js`)],
            },
            {
              test: /\.mdx?$/,
              include: require.resolve(`@yarnpkg/monorepo/CHANGELOG.md`),
              use: [
                utils.getJSLoader({isServer}),
                {
                  loader: require.resolve(`@docusaurus/mdx-loader`),
                  options: {
                    admonitions: true,
                    siteDir: context.siteDir,
                    staticDirs: context.siteConfig.staticDirectories.map(dir => path.resolve(context.siteDir, dir)),
                    isMDXPartial: () => true,
                    isMDXPartialFrontMatterWarningDisabled: true,
                    markdownConfig: context.siteConfig.markdown,
                    ...options.changelog,
                  } satisfies MDXLoaderOptions,
                },
              ],
            },
          ],
        },
        resolve: {
          fallback: {
            fs: false,
            module: false,
            buffer: false,
            os: require.resolve(`os-browserify`),
            path: require.resolve(`path-browserify`),
          },
          alias: {
            '@mdx-js/react': require.resolve(`@mdx-js/react`),
            react: path.resolve(require.resolve(`react/package.json`), `..`),
          },
        },
        plugins: [
          new webpack.DefinePlugin({
            [`process.env`]: JSON.stringify({
              NODE_ENV: `development`,
            }),
            [`process.platform`]: JSON.stringify(`browser`),
            [`process.versions`]: JSON.stringify({
              node: `18.0.0`,
            }),
          }),
        ],
      };
    },
  };
};

// eslint-disable-next-line arca/no-default-export
export default plugin;
