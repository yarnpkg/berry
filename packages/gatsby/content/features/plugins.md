---
category: features
path: /features/plugins
title: "Plugins"
description: An overview of Yarn's plugin capabilities.
---

Ever since Yarn was created, our very essence has been about experimenting, evolving, pushing the status quo - rinse and repeat, each time one step further, each time paving the way for our next move.

As you can guess, this philosophy (coupled with the high number of external contributions we receive) requires us to iterate fast in order to accommodate the various experiments that we brew. In a major step forward, Yarn got redesigned in the v2 in order to leverage a new modular API that can be extended through plugins. Nowadays, most of our features are implemented through those plugins - even `yarn add` and `yarn install` are preinstalled plugins!

```toc
# This code block gets replaced with the Table of Contents
```

## What can plugins do?

  - **Plugins can add new resolvers.** Resolvers are the components tasked from converting dependency ranges (for example `^1.2.0`) into fully-qualified package references (for example `npm:1.2.0`). By implementing a resolver, you can tell Yarn which versions are valid candidates to a specific range.

  - **Plugins can add new fetchers.** Fetchers are the components that take the fully-qualified package references we mentioned in the previous step (for example `npm:1.2.0`) and know how to obtain the data of the package they belong to. Fetchers can work with remote packages (for example the npm registry), but can also find the packages directly from their location on the disk (or any other data source).

  - **Plugins can add new linkers.** Once all the packages have been located and are ready for installation, Yarn will call the linkers to generate the files needed for the install targets to work properly. As an example, the PnP linker would generate the `.pnp.cjs` manifest, and a Python linker would instead generate the virtualenv files needed.

  - **Plugins can add new commands.** Each plugin can ship as many commands as they see fit, which will be injected into our CLI (also making them available through `yarn --help`). Because the Yarn plugins are dynamically linked with the running Yarn process, they can be very small and guaranteed to share the exact same behavior as your package manager (which wouldn't be the case if you were to reimplement the workspace detection, for example).

  - **Plugins can register to some events.** Yarn has a concept known as "hooks", where events are periodically triggered during the lifecycle of the package manager. Plugins can register to those hooks in order to add their own logic depending on what the core allows. For example, the `afterAllInstalled` hook will be called each time the `Project#install` method ends - typically after each `yarn install`.

  - **Plugins can be integrated with each other.** Each plugin has the ability to trigger special actions called hooks, and to register themselves to any defined hook. So for example, you could make a plugin that would execute an action each time a package is added as dependency of one of your workspaces!

## How to write plugins?

We have a tutorial for this! Head over to [Plugin Tutorial](/advanced/plugin-tutorial).

## Official plugins

```
yarn plugin import <name>
```

- [**constraints**](https://github.com/yarnpkg/berry/tree/master/packages/plugin-constraints) - Adds new commands to Yarn to enforce lint rules across workspaces. See the [dedicated page](/features/constraints) for more information.

- [**exec**](https://github.com/yarnpkg/berry/tree/master/packages/plugin-exec) - Adds a new protocol to Yarn (`exec:`) that dynamically generates arbitrary packages rather than downloading them from a known location. See the [plugin page](https://github.com/yarnpkg/berry/tree/master/packages/plugin-exec) for more information.

- [**interactive-tools**](https://github.com/yarnpkg/berry/tree/master/packages/plugin-interactive-tools) - Adds various commands providing a more high-level control on your project using graphical terminal interfaces (for example [`yarn upgrade-interactive`](/cli/upgrade-interactive)).

- [**stage**](https://github.com/yarnpkg/berry/tree/master/packages/plugin-stage) - Adds a new command to Yarn (`yarn stage`) to automatically stage & commit all changes to Yarn-related files in a single line.

- [**typescript**](https://github.com/yarnpkg/berry/tree/master/packages/plugin-typescript) - Improves the TypeScript experience (for example by automatically adding `@types` packages as dependencies when needed). See the [plugin page](https://github.com/yarnpkg/berry/tree/master/packages/plugin-typescript) for more information.

- [**version**](https://github.com/yarnpkg/berry/tree/master/packages/plugin-version) - Adds a new workflow to Yarn (`yarn version`) to efficiently manage releases in a monorepository. See the [dedicated page](/features/release-workflow) for more information.

- [**workspace-tools**](https://github.com/yarnpkg/berry/tree/master/packages/plugin-workspace-tools) - Adds various commands that make working with workspaces a more pleasing experience (for example [`yarn workspaces foreach`](/cli/workspaces/foreach)).

## Contrib plugins

```
yarn plugin import <bundle url>
```

This is just a centralized list of third-party plugins to make discovery easier. No guarantees are made as to plugin quality, compatibility, or lack of malicious code. As with all third-party dependencies, you should review them yourself before including them in your project.

- [**plugin-installs**](https://gitlab.com/Larry1123/yarn-contrib/-/blob/master/packages/plugin-production-install/README.md) by [**Larry1123**](https://gitlab.com/Larry1123) - create minimal yarn installs after removing development dependencies, unrelated workspaces, `@types` packages, etc

- [**yarn.build**](https://yarn.build/) by [**Owen Kelly**](https://github.com/ojkelly/yarn.build) - run builds and tests in parallel according to your workspace dependencies if they have changed, and bundle them up into deployable apps for Docker, AWS Lambda, or any other server.

- [**licenses**](https://github.com/tophat/yarn-plugin-licenses) by [**Noah Negin-Ulster**](https://noahnu.com/) - audit direct and indirect dependency licenses to ensure compliance

- [**semver-up**](https://github.com/tophat/yarn-plugin-semver-up) by [**Noah Negin-Ulster**](https://noahnu.com/) - configurable `yarn up` command that preserves semantic version ranges and update groups

- [**conditions**](https://github.com/nicolo-ribaudo/yarn-plugin-conditions) by [**Nicol√≤ Ribaudo**](https://twitter.com/NicoloRibaudo) - allow choosing between different dependency versions via install-time (during development) and publish-time flags

If you wrote a plugin yourself, feel free to [open a PR](https://github.com/yarnpkg/berry/edit/master/packages/gatsby/content/features/plugins.md) to add it at the end of this list!
