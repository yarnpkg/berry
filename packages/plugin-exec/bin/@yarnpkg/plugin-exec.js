/* eslint-disable */
module.exports = {
name: "@yarnpkg/plugin-exec",
factory: function (require) {
var plugin;plugin=(()=>{"use strict";var e={230:(e,t,r)=>{r.r(t),r.d(t,{execUtils:()=>a,default:()=>p});var a={};r.r(a),r.d(a,{loadGeneratorFile:()=>l,makeLocator:()=>c,makeSpec:()=>i,parseSpec:()=>s});var o=r(966),n=r(688);function s(e){const{params:t,selector:r}=o.structUtils.parseRange(e),a=n.npath.toPortablePath(r);return{parentLocator:t&&"string"==typeof t.locator?o.structUtils.parseLocator(t.locator):null,path:a}}function i({parentLocator:e,path:t,generatorHash:r,protocol:a}){const n=null!==e?{locator:o.structUtils.stringifyLocator(e)}:{},s=void 0!==r?{hash:r}:{};return o.structUtils.makeRange({protocol:a,source:t,selector:t,params:{...s,...n}})}function c(e,{parentLocator:t,path:r,generatorHash:a,protocol:n}){return o.structUtils.makeLocator(e,i({parentLocator:t,path:r,generatorHash:a,protocol:n}))}async function l(e,t,r){const{parentLocator:a,path:s}=o.structUtils.parseFileStyleRange(e,{protocol:t}),i=n.ppath.isAbsolute(s)?{packageFs:new n.CwdFS(n.PortablePath.root),prefixPath:n.PortablePath.dot,localPath:n.PortablePath.root}:await r.fetcher.fetch(a,r),c=i.localPath?{packageFs:new n.CwdFS(n.PortablePath.root),prefixPath:n.ppath.relative(n.PortablePath.root,i.localPath)}:i;i!==c&&i.releaseFs&&i.releaseFs();const l=c.packageFs,p=n.ppath.join(c.prefixPath,s);return await l.readFilePromise(p,"utf8")}const p={fetchers:[class{supports(e,t){return!!e.reference.startsWith("exec:")}getLocalPath(e,t){const{parentLocator:r,path:a}=o.structUtils.parseFileStyleRange(e.reference,{protocol:"exec:"});if(n.ppath.isAbsolute(a))return a;const s=t.fetcher.getLocalPath(r,t);return null===s?null:n.ppath.resolve(s,a)}async fetch(e,t){const r=t.checksums.get(e.locatorHash)||null,[a,n,s]=await t.cache.fetchPackageFromCache(e,r,{onHit:()=>t.report.reportCacheHit(e),onMiss:()=>t.report.reportCacheMiss(e),loader:()=>this.fetchFromDisk(e,t),skipIntegrityCheck:t.skipIntegrityCheck});return{packageFs:a,releaseFs:n,prefixPath:o.structUtils.getIdentVendorPath(e),localPath:this.getLocalPath(e,t),checksum:s}}async fetchFromDisk(e,t){const r=await l(e.reference,"exec:",t);return n.xfs.mktempPromise(async a=>{const s=n.ppath.join(a,"generator.js");return await n.xfs.writeFilePromise(s,r),n.xfs.mktempPromise(async r=>{if(await this.generatePackage(r,e,s,t),!n.xfs.existsSync(n.ppath.join(r,"build")))throw new Error("The script should have generated a build directory");return await o.tgzUtils.makeArchiveFromDirectory(n.ppath.join(r,"build"),{prefixPath:o.structUtils.getIdentVendorPath(e),compressionLevel:t.project.configuration.get("compressionLevel")})})})}async generatePackage(e,t,r,a){return await n.xfs.mktempPromise(async s=>{const i=await o.scriptUtils.makeScriptEnv({project:a.project,binFolder:s}),c=n.ppath.join(e,"runtime.js");return await n.xfs.mktempPromise(async a=>{const s=n.ppath.join(a,"buildfile.log"),l=n.xfs.createWriteStream(s),p=l,u=n.ppath.join(e,"generator"),h=n.ppath.join(e,"build");await n.xfs.mkdirPromise(u),await n.xfs.mkdirPromise(h);const f={tempDir:n.npath.fromPortablePath(u),buildDir:n.npath.fromPortablePath(h),locator:o.structUtils.stringifyLocator(t)};await n.xfs.writeFilePromise(c,`\n          // Expose 'Module' as a global variable\n          Object.defineProperty(global, 'Module', {\n            get: () => require('module'),\n            configurable: true,\n            enumerable: false,\n          });\n\n          // Expose non-hidden built-in modules as global variables\n          for (const name of Module.builtinModules.filter((name) => name !== 'module' && !name.startsWith('_'))) {\n            Object.defineProperty(global, name, {\n              get: () => require(name),\n              configurable: true,\n              enumerable: false,\n            });\n          }\n\n          // Expose the 'execEnv' global variable\n          Object.defineProperty(global, 'execEnv', {\n            value: {\n              ...${JSON.stringify(f)},\n            },\n            enumerable: true,\n          });\n        `);let d=i.NODE_OPTIONS||"";d=d.replace(/\s*--require\s+\S*\.pnp\.c?js\s*/g," ").trim(),i.NODE_OPTIONS=d,l.write(`# This file contains the result of Yarn generating a package (${o.structUtils.stringifyLocator(t)})\n`),l.write("\n");const{code:g}=await o.execUtils.pipevp(process.execPath,["--require",n.npath.fromPortablePath(c),n.npath.fromPortablePath(r),o.structUtils.stringifyIdent(t)],{cwd:e,env:i,stdin:null,stdout:l,stderr:p});if(0!==g)throw n.xfs.detachTemp(a),new Error(`Package generation failed (exit code ${g}, logs can be found here: ${s})`)})})}}],resolvers:[class{supportsDescriptor(e,t){return!!e.range.startsWith("exec:")}supportsLocator(e,t){return!!e.reference.startsWith("exec:")}shouldPersistResolution(e,t){return!1}bindDescriptor(e,t,r){return o.structUtils.bindDescriptor(e,{locator:o.structUtils.stringifyLocator(t)})}getResolutionDependencies(e,t){return[]}async getCandidates(e,t,r){if(!r.fetchOptions)throw new Error("Assertion failed: This resolver cannot be used unless a fetcher is configured");const{path:a,parentLocator:n}=s(e.range);if(null===n)throw new Error("Assertion failed: The descriptor should have been bound");const i=await l(o.structUtils.makeRange({protocol:"exec:",source:a,selector:a,params:{locator:o.structUtils.stringifyLocator(n)}}),"exec:",r.fetchOptions);return[c(e,{parentLocator:n,path:a,generatorHash:o.hashUtils.makeHash("1",i).slice(0,6),protocol:"exec:"})]}async getSatisfying(e,t,r){return null}async resolve(e,t){if(!t.fetchOptions)throw new Error("Assertion failed: This resolver cannot be used unless a fetcher is configured");const r=await t.fetchOptions.fetcher.fetch(e,t.fetchOptions),a=await o.miscUtils.releaseAfterUseAsync(async()=>await o.Manifest.find(r.prefixPath,{baseFs:r.packageFs}),r.releaseFs);return{...e,version:a.version||"0.0.0",languageName:t.project.configuration.get("defaultLanguageName"),linkType:o.LinkType.HARD,dependencies:a.dependencies,peerDependencies:a.peerDependencies,dependenciesMeta:a.dependenciesMeta,peerDependenciesMeta:a.peerDependenciesMeta,bin:a.bin}}}]}},966:e=>{e.exports=require("@yarnpkg/core")},688:e=>{e.exports=require("@yarnpkg/fslib")}},t={};function r(a){if(t[a])return t[a].exports;var o=t[a]={exports:{}};return e[a](o,o.exports,r),o.exports}return r.d=(e,t)=>{for(var a in t)r.o(t,a)&&!r.o(e,a)&&Object.defineProperty(e,a,{enumerable:!0,get:t[a]})},r.o=(e,t)=>Object.prototype.hasOwnProperty.call(e,t),r.r=e=>{"undefined"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(e,"__esModule",{value:!0})},r(230)})();
return plugin;
}
};