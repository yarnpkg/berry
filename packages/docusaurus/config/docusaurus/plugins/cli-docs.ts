import type {Options as MDXLoaderOptions, MDXPlugin}          from '@docusaurus/mdx-loader';
import type {LoadContext, Plugin, PluginContentLoadedActions} from '@docusaurus/types';
import {YarnVersion, miscUtils}                               from '@yarnpkg/core';
import type {BaseContext, Definition}                         from 'clipanion';
import {Cli}                                                  from 'clipanion';
import glob                                                   from 'fast-glob';
import jiti                                                   from 'jiti';
import path                                                   from 'path';

const PLUGIN_NAME = `docusaurus-plugin-yarn-cli-docs`;
const ROUTE_PREFIX = `/cli`;
function dedent(value: string) {
  return value.trim().split(`\n`).map(line => line.trimStart()).join(`\n`);
}

const loader = jiti(__filename, {
  cache: true,
  requireCache: false,
});

const binaries: Array<{ name: string, getCli: () => Promise<Cli<BaseContext>>, watch: Array<string> }> = [
  {
    name: `@yarnpkg/cli`,
    getCli: async () => {
      const {getCli} = loader(`@yarnpkg/cli`) as typeof import('@yarnpkg/cli');
      return getCli();
    },
    watch: [`../plugin-*/sources/commands/**/*.{ts,tsx}`],
  },
  {
    name: `@yarnpkg/builder`,
    getCli: async () => {
      const commands = Object.values(loader(`@yarnpkg/builder`) as typeof import('@yarnpkg/builder'));
      return Cli.from(commands, {binaryName: `yarn builder`});
    },
    watch: [`../yarnpkg-builder/sources/commands/**/*.ts`],
  },
  {
    name: `@yarnpkg/pnpify`,
    getCli: async () => {
      const proxyLoader = jiti(require.resolve(`@yarnpkg/pnpify/package.json`), {
        cache: true,
        requireCache: false,
      });
      const commandPaths = await glob([`./sources/commands/**/*.ts`], {cwd: path.dirname(proxyLoader.resolve(`./package.json`))});

      const commands = commandPaths.map(commandPath => proxyLoader(commandPath).default);
      return Cli.from(commands, {binaryName: `yarn pnpify`});
    },
    watch: [`../yarnpkg-pnpify/sources/commands/**/*.ts`],
  },
  {
    name: `@yarnpkg/sdks`,
    getCli: async () => {
      const {SdkCommand} = loader(`@yarnpkg/sdks`) as typeof import('@yarnpkg/sdks');
      return Cli.from(SdkCommand, {binaryName: `yarn sdks`});
    },
    watch: [`../yarnpkg-sdks/sources/commands/**/*.ts`],
  },
];

export type Options = {
  remarkPlugins: Array<MDXPlugin>;
};

const plugin = async function(context: LoadContext, options: Options): Promise<Plugin<Record<string, Array<Definition>>>> {
  async function createBinaryPages(packageName: string, definitions: Array<Definition>, actions: PluginContentLoadedActions) {
    const pages = await Promise.all(
      definitions
        .sort((a, b) => a.path.localeCompare(b.path))
        .map(definition => createCommandPage(packageName, definition, actions)),
    );

    const index = await createBinaryIndexPage(packageName, definitions, actions);

    return {
      sidebarItem: {
        type: `category`,
        label: packageName,
        href: index.route.path,
        collapsible: false,
        collapsed: false,
        items: pages.map(page => page.sidebarItem),
      },
      routes: [
        index.route,
        ...pages.map(page => page.route),
      ],
    };
  }

  async function createBinaryIndexPage(packageName: string, definitions: Array<Definition>, actions: PluginContentLoadedActions) {
    const shortName = packageName.split(`/`).pop();

    const frontMatter = packageName === `@yarnpkg/cli`
      ? {
        slug: ROUTE_PREFIX,
        title: `CLI Reference`,
        description: `List of commands distributed with Yarn`,
      }
      : {
        slug: `${ROUTE_PREFIX}/${shortName}`,
        title: `${packageName} CLI Reference`,
        description: `List of commands distributed with ${packageName}`,
      };

    const categories = new Map<string, Array<Definition>>();
    for (const definition of definitions)
      miscUtils.getArrayWithDefault(categories, definition.category ?? `General commands`).push(definition);

    const sections = [
      dedent(`
        ---
        ${Object.entries(frontMatter).map(([key, value]) => `${key}: "${value}"`).join(`\n`)}
        ---

        import Link from '@docusaurus/Link';
      `),

      ...Array.from(categories, ([category, definitions]) => dedent(`
        ${categories.size > 1 ? `## ${category}` : ``}

        <table>
          <tbody>
            ${definitions.map(definition => dedent(`
              <tr>
                <td><Link href="${commandSlug(packageName, definition)}">${definition.path}</Link></td>
                <td>${definition.description ? `${definition.description[0].toUpperCase()}${definition.description.slice(1, -1)}.` : ``}</td>
              </tr>
            `)).join(`\n`)}
          </tbody>
        </table>
      `)),
    ];

    const file = await actions.createData(`${frontMatter.slug.slice(1)}.mdx`, sections.filter(section => section !== null).join(`\n\n`));
    await actions.createData(`${frontMatter.slug.slice(1)}.json`, JSON.stringify({
      id: frontMatter.slug.slice(1),
      title: frontMatter.title,
      description: frontMatter.description,
      slug: frontMatter.slug,
      permalink: frontMatter.slug,
      sidebar: `cli`,

      source: file,
      sourceDirName: path.dirname(file),

      draft: false,
      unlisted: false,
      tags: [],
      version: `current`,

      frontMatter,
    }));

    return {
      route: {
        path: frontMatter.slug,
        exact: true,
        component: `@theme/DocItem`,
        modules: {content: file},
        sidebar: `cli`,
      },
    };
  }

  async function createCommandPage(packageName: string, definition: Definition,  actions: PluginContentLoadedActions) {
    const shortName = packageName.split(`/`).pop();
    const slug = commandSlug(packageName, definition);

    const frontMatter = {
      id: slug.slice(1),
      slug,
      title: definition.path,
      description: definition.description ? `${definition.description[0].toUpperCase()}${definition.description.slice(1, -1)}.` : ``,
    };

    const sections = [
      // Frontmatter
      dedent(`
        ---
        ${Object.entries(frontMatter).map(([key, value]) => `${key}: "${value}"`).join(`\n`)}
        ---
      `),

      // Import
      dedent(`
        import {TerminalCode} from '@site/src/components/TerminalCode'
      `),

      // Package Note
      packageName !== `@yarnpkg/cli`
        ? dedent(`
          :::note
          To use this command, you need to use the [\`${packageName}\`](https://github.com/yarnpkg/berry/blob/master/packages/yarnpkg-${shortName}/README.md) package either:
          - By installing it locally using [\`yarn add\`](/cli/add) and running it using [\`yarn run\`](/cli/run)
          - By downloading and running it in a temporary environment using [\`yarn dlx\`](/cli/dlx)
          :::
        `)
        : null,

      // Description
      frontMatter.description
        ? dedent(`
            <div className="subtitle">
            ${frontMatter.description}
            </div>
          `)
        : null,

      // Usage
      dedent(`
        ## Usage

        <TerminalCode command="${definition.usage}" />
      `),

      // Examples
      definition.examples
        ? dedent(`
            ## Examples

            ${definition.examples.map(([description, example]) => dedent(`
              <p>${description}:</p>

              \`\`\`
              ${example}
              \`\`\`
            `)).join(`\n\n`)}
          `)
        : null,

      // Details
      definition.details
        ? dedent(`
            ## Details

            ${definition.details}
          `)
        : null,

      // Options
      definition.options.length > 0
        ? dedent(`
            ## Options

            | Definition | Description |
            | ---------- | ----------- |
            ${definition.options.map(({definition, description}) => dedent(`
              | <h4 id="${encodeURIComponent((`options-${definition}`).replace(/-+/g, `-`))}" className="header-code"><code className="language-text">${definition}</code></h4> | ${description} |
            `)).join(`\n`)}
          `)
        : null,
    ];

    const file = await actions.createData(`${frontMatter.id}.mdx`, sections.filter(section => section !== null).join(`\n\n`));
    await actions.createData(`${frontMatter.id}.json`, JSON.stringify({
      id: frontMatter.id,
      title: frontMatter.title,
      description: frontMatter.description,
      slug: frontMatter.slug,
      permalink: frontMatter.slug,
      sidebar: `cli`,

      source: file,
      sourceDirName: path.dirname(file),

      draft: false,
      unlisted: false,
      tags: [],
      version: `current`,

      frontMatter,
    }));

    return {
      sidebarItem: {
        type: `link`,
        label: frontMatter.title,
        href: frontMatter.slug,
        docId: frontMatter.id,
        unlisted: false,
      },
      route: {
        path: frontMatter.slug,
        exact: true,
        component: `@theme/DocItem`,
        modules: {content: file},
        sidebar: `cli`,
      },
    };
  }

  function commandSlug(packageName: string, definition: Definition) {
    const route = definition.path.split(` `).slice(1).join(`/`);
    return `${ROUTE_PREFIX}/${route}${packageName === `@yarnpkg/${route}` ? `/default` : ``}`;
  }

  return {
    name: PLUGIN_NAME,
    async loadContent() {
      return Object.fromEntries(
        await Promise.all(binaries.map(async ({name, getCli}) => {
          const cli = await getCli();
          return [name, cli.definitions()] as const;
        })),
      );
    },
    async contentLoaded({content, actions}) {
      const pages = await Promise.all(
        Object.entries(content).map(([packageName, definitions]) => createBinaryPages(packageName, definitions, actions)),
      );

      const version = await actions.createData(`version-metadata.json`, JSON.stringify({
        pluginId: `default`,
        version: `current`,
        label: `master (${YarnVersion})`,
        banner: null,
        badge: false,
        noIndex: false,
        className: `docs-version-current`,
        isLast: true,
        docsSidebars: {
          cli: pages.map(page => page.sidebarItem),
        },
        docs: {},
      }));

      actions.addRoute({
        path: ROUTE_PREFIX,
        exact: false,
        component: `@theme/DocsRoot`,
        routes: [
          {
            path: ROUTE_PREFIX,
            exact: false,
            component: `@theme/DocVersionRoot`,
            modules: {version},
            routes: [
              {
                path: ROUTE_PREFIX,
                exact: false,
                component: `@theme/DocRoot`,
                routes: pages.flatMap(page => page.routes),
              },
            ],
          },
        ],
      });
    },
    configureWebpack(config, isServer, utils) {
      return {
        module: {
          rules: [
            {
              test: /\.mdx$/,
              include: path.join(context.generatedFilesDir, PLUGIN_NAME, `default`),
              use: [
                utils.getJSLoader({isServer}),
                {
                  loader: require.resolve(`@docusaurus/mdx-loader`),
                  options: {
                    admonitions: true,
                    siteDir: context.siteDir,
                    staticDirs: [...context.siteConfig.staticDirectories, path.join(context.siteDir, `static`)],
                    isMDXPartial: () => false,
                    createAssets: () => ({image: undefined}),
                    metadataPath: mdxPath => mdxPath.replace(/.mdx$/, `.json`),
                    remarkPlugins: options.remarkPlugins,
                    markdownConfig: context.siteConfig.markdown,
                  } satisfies MDXLoaderOptions,
                },
              ],
            },
          ],
        },
      };
    },

    getPathsToWatch() {
      return binaries.flatMap(({watch}) => watch);
    },
  };
};

// eslint-disable-next-line arca/no-default-export
export default plugin;
