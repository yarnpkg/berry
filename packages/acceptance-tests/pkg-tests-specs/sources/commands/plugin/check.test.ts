import { hashUtils } from "@yarnpkg/core";
import { xfs, PortablePath } from "@yarnpkg/fslib";

import { createMockPlugin, mockPluginServer } from "../../features/plugins.utility";

describe(`Commands`, () => {
  describe(`plugin check`, () => {
    test(
      `it should run successfully and do nothing`,
      makeTemporaryEnv({}, async ({ path, run, source }) => {
        await mockPluginServer(async (mockServer) => {
          const mockPluginPath = await createMockPlugin(path);
          const { pluginUrl, httpsCaFilePath } = mockServer;

          await xfs.writeJsonPromise(`${path}/.yarnrc.yml` as PortablePath, {
            httpsCaFilePath,
            plugins: [
              {
                path: mockPluginPath,
                spec: pluginUrl,
                checksum: await hashUtils.checksumFile(`${path}/${mockPluginPath}` as PortablePath),
              },
            ],
          });

          await expect(run(`plugin`, `check`)).resolves.toMatchSnapshot();
        });
      }),
    );

    test(
      `it should print out the plugin when it has a different checksum`,
      makeTemporaryEnv({}, async ({ path, run, source }) => {
        await mockPluginServer(async (mockServer) => {
          const mockPluginPath = await createMockPlugin(path);
          const { pluginUrl, httpsCaFilePath } = mockServer;

          await xfs.writeJsonPromise(`${path}/.yarnrc.yml` as PortablePath, {
            httpsCaFilePath,
            plugins: [
              {
                path: mockPluginPath,
                spec: pluginUrl,
                checksum: `I am wrong checksum 123456`,
              },
            ],
          });

          await expect(run(`plugin`, `check`)).rejects.toThrow(/is different/);
        });
      }),
    );
  });
});
