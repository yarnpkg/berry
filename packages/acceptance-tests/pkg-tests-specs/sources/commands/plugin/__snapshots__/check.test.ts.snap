// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Commands plugin check it should run successfully and do nothing 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "âž¤ YN0000: Done
",
}
`;
